streaming/src/data_reader.cc

```diff
@@ -100,106 +105,250 @@ StreamingStatus DataReader::InitChannelMerger() {
+    is_valid_break = true;
+  } else if (timer_interval_ != -1 && cur_time - last_message_ts_ >= timer_interval_ &&
+             message->meta->IsEmptyMsg()) {
+    // sent empty message when reaching timer_interval
     last_message_ts_ = cur_time;
     is_valid_break = true;
   }
 
   offset_info.current_message_id = message->meta->GetLastMessageId();
-  offset_info.current_seq_id = message->seq_id;
   last_bundle_ts_ = message->meta->GetMessageBundleTs();
 
-  STREAMING_LOG(DEBUG) << "[Reader] [Bundle] message type =>"
-                       << static_cast<int>(message->meta->GetBundleType())
-                       << " from id => " << message->from << ", queue seq id =>"
-                       << message->seq_id << ", message id => "
-                       << message->meta->GetLastMessageId();
+  STREAMING_LOG(DEBUG) << "[Reader] [Bundle] Get merged message bundle="
+   << *message << ", is_valid_break=" << is_valid_break;
+  last_fetched_queue_item_ = message;
   return StreamingStatus::OK;
 }
 
+bool DataReader::BarrierAlign(std::shared_ptr<DataBundle> &message) {
+  // Arrange barrier action when barrier is arriving.
+  StreamingBarrierHeader barrier_header;
+  StreamingMessage::GetBarrierIdFromRawData(
+      message->data + kMessageHeaderSize, &barrier_header);
+  uint64_t barrier_id = barrier_header.barrier_id;
+  auto *barrier_align_cnt = &global_barrier_cnt_;
+  auto &channel_info = channel_info_map_[message->from];
+  // Target count is input vector size (global barrier)
+  // or barrier collection length (partial barrier).
+  uint32_t target_count = 0;
+
+  channel_info.barrier_id = barrier_header.barrier_id;
+  target_count = input_queue_ids_.size();
+  (*barrier_align_cnt)[barrier_id]++;
+  // the next message checkpoint is changed if this's barrier message.
+  STREAMING_LOG(INFO) << "[Reader] [Barrier] get barrier, barrier_id=" << barrier_id
+                      << ", barrier_cnt=" << (*barrier_align_cnt)[barrier_id]
+                      << ", global barrier id=" << barrier_header.barrier_id
+                      << ", from q_id=" << message->from << ", barrier type ="
+                      << static_cast<uint32_t>(barrier_header.barrier_type)
+                      << ", target count =" << target_count;
+  // notify invoker the last barrier, so that checkpoint or something related can be
+  // taken right now.
+  if ((*barrier_align_cnt)[barrier_id] == target_count) {
+    // map can't be used in multithread (crash in report timer)
+    barrier_align_cnt->erase(barrier_id);
+    STREAMING_LOG(INFO)
+        << "[Reader] [Barrier] last barrier received, return barrier. barrier_id = "
+        << barrier_id << ", from q_id=" << message->from;
+    return true;
+  }
+  return false;
+}
+
 StreamingStatus DataReader::GetBundle(const uint32_t timeout_ms,
                                       std::shared_ptr<DataBundle> &message) {
+  STREAMING_LOG(DEBUG) << "GetBundle, timeout_ms=" << timeout_ms;
```

Comment:
This log doesn't make much sense, maybe can be removed.

Reply:
I think we should keep it, it's useful when debugging.